ambry-commons/src/test/java/com.github.ambry.commons/HelixNotifierTest.java

```diff
@@ -0,0 +1,365 @@
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.helix.ZNRecord;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import static com.github.ambry.commons.TestUtils.*;
+import static org.junit.Assert.*;
+
+
+/**
+ * Unit tests for {@link HelixNotifier}.
+ */
+public class HelixNotifierTest {
+  private static int zkClientConnectTimeoutMs = 20000;
+  private static int zkClientSessionTimeoutMs = 20000;
+  private static String zkClientConnectString = "localhost:2182";
+  private static String storeRootPath = "/ambry/testCluster/helixPropertyStore";
+
+  private static final HelixPropertyStoreFactory<ZNRecord> storeFactory = new MockHelixPropertyStoreFactory<>();
+  private static final List<String> refTopics = Arrays.asList(new String[]{"Topic1", "Topic2"});
+  private static final List<String> refMessages = Arrays.asList(new String[]{"Message1", "Message2"});
+  private final List<String> receivedTopicsByListener0 = new ArrayList<>();
+  private final List<String> receivedTopicsByListener1 = new ArrayList<>();
+  private final List<String> receivedMessagesByListener0 = new ArrayList<>();
+  private final List<String> receivedMessagesByListener1 = new ArrayList<>();
+  private final AtomicReference<CountDownLatch> latch0 = new AtomicReference<>();
+  private final AtomicReference<CountDownLatch> latch1 = new AtomicReference<>();
+  private static HelixPropertyStoreConfig storeConfig;
+  private static long latchTimeoutMs = 1000;
+  private HelixNotifier notifier;
+  private List<TopicListener<String>> listeners;
+
+  @BeforeClass
+  public static void initialize() throws Exception {
+    storeConfig = HelixPropertyStoreUtils.getHelixStoreConfig(zkClientConnectString, zkClientSessionTimeoutMs,
+        zkClientConnectTimeoutMs, storeRootPath);
+  }
+
+  @Before
+  public void setup() throws Exception {
+    resetTopicListeners();
+    HelixPropertyStoreUtils.deleteStoreIfExists(storeConfig, storeFactory);
+    listeners = new ArrayList<>();
+    notifier = new HelixNotifier(storeConfig, storeFactory);
+    listeners.add(new ListenerForTest(latch0, receivedTopicsByListener0, receivedMessagesByListener0));
+    listeners.add(new ListenerForTest(latch1, receivedTopicsByListener1, receivedMessagesByListener1));
+  }
+
+  @AfterClass
+  public static void cleanStore() throws Exception {
+    HelixPropertyStoreUtils.deleteStoreIfExists(storeConfig, storeFactory);
+  }
+
+  /**
+   * Tests normal operations using a single {@link HelixNotifier} and two {@link TopicListener}s.
+   * @throws Exception Any unexpected exceptions.
+   */
+  @Test
+  public void testHelixNotifier() throws Exception {
+    // Subscribe a topic and send a message for the topic
+    notifier.subscribe(refTopics.get(0), listeners.get(0));
+    latch0.set(new CountDownLatch(1));
+    notifier.sendMessage(refTopics.get(0), refMessages.get(0));
+    awaitLatchOrTimeout(latch0.get(), latchTimeoutMs);
```

Comment:
Do you think it makes sense to also ensure that a notification comes in only once? `countDownLatch.countDown()` does not throw any error if the count is already zero.